added the max_matching_ngram_size to GenerationConfig

mosheber opened the pull request:
# What does this PR do?

* Added the max_matching_ngram_size parameter into the GenerationConfig, for the PromptLookupCandidateGenerator.
* Included the max_matching_ngram_size when calling the __init__ of PromptLookupCandidateGenerator in   _get_candidate_generator, in case it is specified. 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante , would appreciate it if you could give this PR a glance, and thank you in advance.

mosheber committed:
added the max_matching_ngram_size parameter into the GenerationConfig, for the PromptLookupCandidateGenerator

gante commented:
Hi @mosheber 👋 

I'd be happy to merge the PR, conditional on the answer to the following question being yes (preferably backed with data): have you found significant benefits of changing the flag you added?

On the [original issue](https://github.com/huggingface/transformers/issues/27722), the author's experiments showed that there were little benefits in changing this option. As such, we don't want to add new flags unless they result in clear benefits :)

danielkorat commented:
Hi @gante @mosheber 
The results below show a `3ms` speedup in latency with a 7B target model, when comparing the default `max_matching_ngram_size=2` with `max_matching_ngram_size=4`.
**As the target size increases further, the speedup will be greater.**
Note that these results use my optimized routine for PLD subsequence matching (~`70x` faster on avg, uses `numba`): 
```bash
get_candidates_orig: 0.3467 ms
get_candidates:      0.0047 ms
```

![Screenshot 2024-02-26 at 18 17 31](https://github.com/huggingface/transformers/assets/32893314/89cb9242-baab-4282-9165-1fb2aa72522e)
![Screenshot 2024-02-26 at 18 17 11](https://github.com/huggingface/transformers/assets/32893314/ac83f2a6-c198-45f3-b1d7-08edc7dde9e0)

gante commented:
@danielkorat convinced by your numbers 👍 

Let's add this PR!

gante commented on src/transformers/generation/utils.py:
```diff
@@ -698,9 +698,12 @@ def _get_candidate_generator(
         """
         if generation_config.prompt_lookup_num_tokens is not None:
-            candidate_generator = PromptLookupCandidateGenerator(
+            candidate_generator_params = dict(
```
Instead of creating a dict here, let's:
a) pass keyword arguments to `PromptLookupCandidateGenerator` (as before the PR)
b) default `max_matching_ngram_size` in `PromptLookupCandidateGenerator` to `None`, and set it to the original default value in `__init__` if it is `None`.

This pushes complexity away from `generate` and into `PromptLookupCandidateGenerator` :)

mosheber replied:
Thanks for the comment! I switched back to keyword arguments.

gante commented:
@mosheber after applying the fix, please run `make fixup` from the `transformers` folder and commit the result. It will fix the CI issues you're seeing :)

mosheber committed:
switched back to keyword arguments

mosheber committed:
Merge branch 'huggingface:main' into main

mosheber committed:
Merge branch 'main' of https://github.com/mosheber/transformers

mosheber commented:
> @mosheber after applying the fix, please run `make fixup` from the `transformers` folder and commit the result. It will fix the CI issues you're seeing :)

I ran the `make fixup` as well.

gante reviewed (approved):
The changes look good to me, thank you for iterating and making `transformers` better 💛

gante reviewed (commented):
@mosheber 

Actually, there is a tiny missing thing: `max_matching_ngram_size` should have an entry in the docstring of `GenerationConfig`

our CI is still complaining about code formating, make sure you have the latest version installed when running `make fixup` again :)

mosheber committed:
added PromptLookupCandidateGenerator docstring for its parameters

mosheber committed:
ruff reformat

mosheber commented:
> @mosheber
> 
> Actually, there is a tiny missing thing: `max_matching_ngram_size` should have an entry in the docstring of `GenerationConfig`
> 
> our CI is still complaining about code formating, make sure you have the latest version installed when running `make fixup` again :)

Great idea! I added the doc string to the GenerationConfig class. 
Regarding the tests, I fixed the formatting issue, it seems that the CI fails due to timing reasons, which are not related to this PR.

ArthurZucker commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -257,11 +257,18 @@ class GenerationConfig(PushToHubMixin):
+        prompt_lookup_num_tokens (`int`, *optional*, default to `None`):
+            The number of tokens to be output as candidate tokens.
+
+        max_matching_ngram_size (`int`, *optional*, default to `None`):
+            The maximum ngram size to be considered for matching in the prompt
```
```suggestion
        max_matching_ngram_size (`int`, *optional*, default to `None`):
            The maximum ngram size to be considered for matching in the prompt. Default to 2 if not provided.
```

ArthurZucker reviewed (approved):
LGTM! just one nit

gante committed:
Update src/transformers/generation/configuration_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29131). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
